core-compose/src/main/java/com/squareup/workflow/ui/compose/WorkflowContainer.kt

```diff
@@ -0,0 +1,323 @@
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param viewEnvironment The [ViewEnvironment] used to show the [ComposeRendering]s emitted by
+ * the workflow.
+ * @param props The props to render the root workflow with. If this value changes between calls,
+ * the workflow runtime will re-render with the new props.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ */
+@Direct
+@Composable fun <PropsT> WorkflowContainer(
+  workflow: Workflow<PropsT, Nothing, ComposeRendering>,
+  viewEnvironment: ViewEnvironment,
+  props: PropsT,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, viewEnvironment, props, {}, modifier, diagnosticListener)
+}
+
+/**
+ * Render a [Workflow]'s renderings.
+ *
+ * When this function is first composed it will start a new runtime. This runtime will be restarted
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param viewEnvironment The [ViewEnvironment] used to show the [ComposeRendering]s emitted by
+ * the workflow.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ */
+@Direct
+@Composable fun WorkflowContainer(
+  workflow: Workflow<Unit, Nothing, ComposeRendering>,
+  viewEnvironment: ViewEnvironment,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, viewEnvironment, Unit, {}, modifier, diagnosticListener)
+}
+
+@Composable private fun <PropsT, OutputT : Any, RenderingT> observeWorkflow(
+  @Pivotal workflow: Workflow<PropsT, OutputT, RenderingT>,
+  props: PropsT,
+  onOutput: (OutputT) -> Unit,
+  @Pivotal coroutineContext: CoroutineContext,
+  @Pivotal diagnosticListener: WorkflowDiagnosticListener? = null
+): RenderingT {
+  // This can be a StateFlow once coroutines is upgraded to 1.3.6.
+  val propsChannel = remember { Channel<PropsT>(capacity = CONFLATED) }
+  propsChannel.offer(props)
+
+  // Need a mutable holder for onOutput so the outputs subscriber created in the onActive block
+  // will always be able to see the latest value.
+  val outputCallback = remember { OutputCallback(onOutput) }
+  outputCallback.onOutput = onOutput
+
+  val renderingState = state<RenderingT?> { null }
+  val snapshotState = savedInstanceState(saver = SnapshotSaver) { null }
+
+  // We can't use onActive/on(Pre)Commit because they won't run their callback until after this
+  // function returns, and we need to run this immediately so we get the rendering synchronously.
+  val workflowScope = remember {
```

Comment:
This might actually have hidden bugs. The core compose support for livedata & rxjava just moved away from `remember` to `onPreCommit` because of some edge cases ([CL](https://android.googlesource.com/platform/frameworks/support/+/f1cb087e3cabfc0ed91bd96825b6a89209a3a66e)).

Reply:
I'm going to leave as-is for now, until I'm sure there's actually a problem.